src/DDTrace/Util/ContainerInfo.php

```diff
@@ -0,0 +1,50 @@
+<?php
+
+namespace DDTrace\Util;
+
+/**
+ * Utility class to extract container info.
+ */
+class ContainerInfo
+{
+    private $cgroupProcFile;
+
+    // Example Docker
+    // 13:name=systemd:/docker/3726184226f5d3147c25fdeab5b60097e378e8a720503a5e19ecfdf29f869860
+    // Example Kubernetes
+    // 11:perf_event:/kubepods/something/pod3d274242-8ee0-11e9-a8a6-1e68d864ef1a/3e74d3fd9db4c9dd921ae05c2502fb984d0cde1b36e581b13f79c639da4518a1
+    // Example ECS
+    // 9:perf_event:/ecs/user-ecs-classic/5a0d5ceddf6c44c1928d367a815d890f/38fac3e99302b3622be089dd41e7ccf38aff368a86cc339972075136ee2710ce
+    // Example Fargate
+    // 11:something:/ecs/5a081c13-b8cf-4801-b427-f4601742204d/432624d2150b349fe35ba397284dea788c2bf66b885d14dfc1569b01890ca7da
+    const LINE_RE = '/^(\d+):([^:]*):.*\/([0-9a-f]{64})$/';
+
+    public function __construct($cgroupProcFile = '/proc/self/cgroup')
+    {
+        $this->cgroupProcFile = $cgroupProcFile;
+    }
+
+    /**
+     * Extracts the container id if the application runs in a containerized environment, `null` otherwise.
+     * Note that the value is not cached, so invoking this method multiple times might lead to performance
+     * degradation as one IO operation and possibly a few regex match operations are required.
+     *
+     * @return string|null
+     */
+    public function getContainerId()
+    {
+        if (!file_exists($this->cgroupProcFile)) {
+            return null;
+        }
+
+        $lines = file($this->cgroupProcFile);
```

Comment:
what will happen if the file is not user readable ?

Reply:
Also What will happen if file is 100MBs long ?

Reply:
Very good points, both:

> what will happen if the file is not user readable ?

Let me add a try-catch block.

> Also What will happen if file is 100MBs long ?

Not sure this is very realistic for this file specifically, but still we just a little bit of complexity we can sleep much better :) Are you cool with using [`fgets`](https://www.php.net/manual/en/function.fgets.php)? We now break if one single line is 100MB instead of the while file.

Again we are being extra-defensive here but I agree with that. On the other hand, probably going beyond this (unless you have some other idea) might be possibly overkill.

Reply:
`fgets` looks good!
Optionally we could add a fgets read limit. But I think current implementation already might be enough since if that file is corrupted then it means the whole system is completely broken. 

One thing missing I think - would be to add some test to check what happens if we access a read protected file - if possible

Reply:
> One thing missing I think - would be to add some test to check what happens if we access a read protected file - if possible

Very good point, it might take a while to figure it out how btw :)

Reply:
@pawelchcki added test `testFileNoReadPermissions`